app/bookmark.py

```diff
@@ -0,0 +1,115 @@
+# Copyright 2018 Google Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+class Bookmark(object):
+  """
+  This bookmark object is used as a marker for different places in a
+  text document. Note that because text buffer lines index at 0, all
+  references to rows also assume that 0 is the first line.
```

Comment:
Trivia: that's part of the reason I use the term 'row' rather than 'line number'. I use row (or index) for a zero based value; and 'line' for a one based value. i.e. (row = line - 1), which I hope is true throughout the code (please let me know if we're being inconsistent somewhere).